File: adapters/mcp-ts/src/translation.ts
/**
 * ANIP → MCP translation layer.
 *
 * Converts ANIP capability declarations into MCP tool schemas,
 * enriching descriptions with ANIP metadata that MCP cannot
 * natively represent.
 */

import type { ANIPCapability } from "./discovery.js";

const TYPE_MAP: Record<string, string> = {
  string: "string",
  integer: "integer",
  number: "number",
  boolean: "boolean",
  date: "string",
  airport_code: "string",
};

export interface MCPInputSchema {
  type: "object";
  properties: Record<string, Record<string, unknown>>;
  required?: string[];
  [key: string]: unknown;
}

export function capabilityToInputSchema(
  capability: ANIPCapability
): MCPInputSchema {
  const properties: Record<string, Record<string, unknown>> = {};
  const required: string[] = [];

  for (const input of capability.inputs) {
    const jsonType = TYPE_MAP[input.type] ?? "string";
    const prop: Record<string, unknown> = {
      type: jsonType,
      description: input.description ?? "",
    };

    if (input.type === "date") {
      prop.format = "date";
    }
    if (input.default !== undefined && input.default !== null) {
      prop.default = input.default;
    }

    properties[input.name] = prop;

    if (input.required !== false) {
      required.push(input.name);
    }
  }

  const schema: MCPInputSchema = { type: "object", properties };
  if (required.length > 0) {
    schema.required = required;
  }
  return schema;
}

export function enrichDescription(capability: ANIPCapability): string {
  const parts: string[] = [capability.description];

  // Side-effect warning
  if (capability.sideEffect === "irreversible") {
    parts.push("WARNING: IRREVERSIBLE action — cannot be undone.");
    if (capability.rollbackWindow === "none") {
      parts.push("No rollback window.");
    }
  } else if (capability.sideEffect === "write") {
    const rollback = capability.rollbackWindow;
    if (rollback && rollback !== "none" && rollback !== "not_applicable") {
      parts.push(`Reversible within ${rollback}.`);
    }
  } else if (capability.sideEffect === "read") {
    parts.push("Read-only, no side effects.");
  }

  // Financial cost
  if (capability.financial && capability.cost) {
    const financial = capability.cost.financial as Record<string, unknown> | undefined;
    const certainty = capability.cost.certainty as string | undefined;

    if (certainty === "fixed" && financial) {
      const amount = financial.amount as number;
      const currency = (financial.currency as string) ?? "USD";
      if (amount > 0) {
        parts.push(`Cost: ${currency} ${amount} (fixed).`);
      }
    } else if (certainty === "estimated" && financial) {
      const rangeMin = financial.range_min as number | undefined;
      const rangeMax = financial.range_max as number | undefined;
      const currency = (financial.currency as string) ?? "USD";
      if (rangeMin !== undefined && rangeMax !== undefined) {
        parts.push(`Estimated cost: ${currency} ${rangeMin}-${rangeMax}.`);
      }
    } else if (certainty === "dynamic" && financial) {
      const upperBound = financial.upper_bound as number | undefined;
      const currency = (financial.currency as string) ?? "USD";
      if (upperBound !== undefined) {
        parts.push(`Dynamic cost, up to ${currency} ${upperBound}.`);
      } else {
        parts.push("Dynamic cost — amount varies.");
      }
    }
  }

  // Prerequisites
  if (capability.requires.length > 0) {
    const prereqs = capability.requires.map((r) => r.capability);
    parts.push(`Requires calling first: ${prereqs.join(", ")}.`);
  }

  // Scope requirements
  if (capability.minimumScope.length > 0) {
    parts.push(`Delegation scope: ${capability.minimumScope.join(", ")}.`);
  }

  return parts.join(" ");
}

File: bridges/mcp/anip_mcp_bridge/translation.py
"""ANIP → MCP translation layer.

Converts ANIP capability declarations into MCP tool schemas,
enriching descriptions with ANIP metadata that MCP cannot
natively represent.
"""

from __future__ import annotations

from .discovery import ANIPCapability

# Map ANIP input types to JSON Schema types
_TYPE_MAP = {
    "string": "string",
    "integer": "integer",
    "number": "number",
    "boolean": "boolean",
    "date": "string",  # JSON Schema has no date type; use string with format
    "airport_code": "string",
}


def capability_to_input_schema(capability: ANIPCapability) -> dict:
    """Convert ANIP capability inputs to JSON Schema for MCP tool."""
    properties = {}
    required = []

    for inp in capability.inputs:
        json_type = _TYPE_MAP.get(inp.get("type", "string"), "string")
        prop: dict = {"type": json_type, "description": inp.get("description", "")}

        # Add format hint for date types
        if inp.get("type") == "date":
            prop["format"] = "date"

        # Add default if specified
        if "default" in inp and inp["default"] is not None:
            prop["default"] = inp["default"]

        properties[inp["name"]] = prop

        if inp.get("required", True):
            required.append(inp["name"])

    schema: dict = {"type": "object", "properties": properties}
    if required:
        schema["required"] = required

    return schema


def enrich_description(capability: ANIPCapability) -> str:
    """Enrich MCP tool description with ANIP metadata.

    Since MCP has no structured fields for side-effects, costs,
    or prerequisites, we encode these signals into the description
    string. This is lossy but gives MCP-native agents safety hints
    they would otherwise lack entirely.
    """
    parts = [capability.description]

    # Side-effect warning
    if capability.side_effect == "irreversible":
        parts.append(
            "WARNING: IRREVERSIBLE action — cannot be undone."
        )
        if capability.rollback_window == "none":
            parts.append("No rollback window.")
    elif capability.side_effect == "write":
        rollback = capability.rollback_window
        if rollback and rollback not in ("none", "not_applicable"):
            parts.append(f"Reversible within {rollback}.")
    elif capability.side_effect == "read":
        parts.append("Read-only, no side effects.")

    # Financial cost
    if capability.financial and capability.cost:
        financial = capability.cost.get("financial", {})
        certainty = capability.cost.get("certainty", "unknown")

        if certainty == "fixed":
            amount = financial.get("amount", "unknown")
            currency = financial.get("currency", "USD")
            if amount and float(amount) > 0:
                parts.append(f"Cost: {currency} {amount} (fixed).")
        elif certainty == "estimated":
            range_min = financial.get("range_min")
            range_max = financial.get("range_max")
            currency = financial.get("currency", "USD")
            if range_min is not None and range_max is not None:
                parts.append(
                    f"Estimated cost: {currency} {range_min}-{range_max}."
                )
        elif certainty == "dynamic":
            upper = financial.get("upper_bound")
            currency = financial.get("currency", "USD")
            if upper is not None:
                parts.append(f"Dynamic cost, up to {currency} {upper}.")
            else:
                parts.append("Dynamic cost — amount varies.")

    # Prerequisites
    if capability.requires:
        prereq_names = [r.get("capability", r) for r in capability.requires]
        parts.append(f"Requires calling first: {', '.join(prereq_names)}.")

    # Scope requirements
    if capability.minimum_scope:
        parts.append(
            f"Delegation scope: {', '.join(capability.minimum_scope)}."
        )

    return " ".join(parts)
